Mathlib/Computability/PrimrecFilter.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2025 David J. Webb. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David J. Webb
+-/
+import Mathlib.Computability.Primrec
+/-!
+
+# Bounded quantifiers are primitive recursive
+This file contains lemmata for showing bounded existentially and universally quantified
+statements are primitive recursive, as well as more general filtering for arbitrary
+primcodable types.
+
+## Main results
+- Filtering for elements from a list that meet a primrec criterion is primrec
+- Checking whether a list has some element meeting a primrec criterion is primrec
+- Checking whether every element of a list meets a primrec criterion is primrec
+- Primitive recursive functions are closed under bounded existential and universal quantifiers
+
+## References
+- [R. I. Soare *Turing Computability - Theory and Applications*] [Soare2016]
+-/
+
+open List Primrec
+
+namespace Primrec
+
+variable {α} [Primcodable α] (f : α → Prop) [DecidablePred f]
+
+/-- Filtering a list for elements that satisfy a decidable predicate is primitive recursive. -/
+lemma list_filter (hf : PrimrecPred f) : Primrec fun L ↦ (filter (fun a ↦ f a) L) := by
+  rw [← List.filterMap_eq_filter]
+  apply listFilterMap Primrec.id
+  simp only [Primrec₂, Option.guard, decide_eq_true_eq]
+  exact ite (PrimrecPred.comp hf snd) (option_some_iff.mpr snd) (const none)
+
+/-- Checking if any element of a list satisfies a decidable predicate is primitive recursive. -/
+lemma filter_exists (hf : PrimrecPred f) : PrimrecPred fun (L : List α) ↦ (∃ a ∈ L, f a) := by
+  let g := fun L ↦ List.filter (fun a ↦ f a) L
+  have h (L : List α): ((g L).length ≠ 0) ↔ (∃ a ∈ L, f a) := by simp [g]
+  apply PrimrecPred.of_eq (PrimrecPred.not ?_) h
+  exact PrimrecRel.comp Primrec.eq (comp list_length (list_filter f hf)) (const 0)
+
+/-- Checking if every element of a list satisfies a decidable predicate is primitive recursive. -/
+lemma filter_forall (hf : PrimrecPred f) : PrimrecPred fun (L : List α) ↦ (∀ a ∈ L, f a) := by
+  let g := fun L ↦ List.filter (fun a ↦ f a) L
+  have h (L : List α): ((g L).length = L.length) ↔ (∀ a ∈ L, f a) := by simp [g]
+  apply PrimrecPred.of_eq ?_ h
+  exact PrimrecRel.comp Primrec.eq (comp list_length (list_filter f hf)) list_length
+
+variable (f : ℕ → Prop) [DecidablePred f]
+
+/-- Bounded existential quantifiers are primitive recursive. -/
+lemma bounded_exists (hf : PrimrecPred f) : PrimrecPred fun n ↦ ∃ x < n, f x :=
+  PrimrecPred.of_eq (PrimrecPred.comp (filter_exists f hf) list_range) (by simp)
+
+/-- Bounded universal quantifiers are primitive recursive. -/
+lemma bounded_forall (hf : PrimrecPred f) : PrimrecPred fun n ↦ ∀ x < n, f x :=
+  PrimrecPred.of_eq (PrimrecPred.comp (filter_forall f hf) list_range) (by simp)
+
+end Primrec
+
+namespace primrec₂
+
+/-- If f a b is decidable, then given L : List α and b : β, it is primitive recurisve
+to filter L for elements a with f a b -/
+
```

Comment:
This doc-string displays a bit nicer in the generated documentation is you write it like this:
```suggestion
/-- If `f a b` is decidable, then given `L : List α` and `b : β`, it is primitive recurisve
to filter `L` for elements `a` with `f a b`. -/